insights/lambdas/update_athena_metric_tables/lambda_update_athena_metric_tables.py

```diff
@@ -0,0 +1,184 @@
+import boto3
+
+from utils.utils import (
+    get_report_dates_from_target_date,
+    check_table_exists,
+    check_table_for_report_date_entry,
+    update_or_create_table_for_report_date,
+)
+
+"""
+Summary:
+
+This lambda supports the updating (or creation) of metric
+tables via Athena used for BB2 Insights dashboards in QuickSight.
+
+It does the following:
+
+- Accepts a lambda parameters dictionary in the following format:
+    {
+      "REGION": "<AWS region>",
+      "WORKGROUP": "bb2",
+      "DATABASE": "bb2",
+      "ENV": "<prod/impl/test>"
+      "BASENAME_MAIN": "<basename of main table>" Ex: "global_state"
+      "BASENAME_PER_APP": "<basename of per-app table>" Ex: "global_state_per_app"
+      "TARGET_DATE": <target report date week> EX: "2022-09-19",
+    }
+
+- Computes the report_date, start_date, and end_date based on the TARGET_DATE param.
+
+- Updates (or creates) the per-applications table for the target report_date,
+  ENV and BASENAME_PER_APP params.
+
+- Updates (or creates) the main (top-level) table for the target report_date,
+  ENV and BASENAME_MAIN params.
+
+For each of the tables, it does the following:
+
+- Check if the table already exists. Setup to create it, if not.
+
+- Check if an entry for the report_date already exists to prevent duplicate
+  entries.
+
+- Execute the relatate SQL from the corresponding template file and
+  update (or create) the table with the results.
+
+- Retry running the SQL if there is a time-out up to 3-times.
+
+"""
+
+
+TEMPLATE_FILE_PER_APP = (
+    "./sql_templates/template_generate_per_app_metrics_for_report_date.sql"
+)
+
+TEMPLATE_FILE_MAIN = "./sql_templates/template_generate_metrics_for_report_date.sql"
+
+
+def update_or_create_metrics_table(session, params, table_basename, template_file):
+
+    print("##")
+    print(
+        "## --- UPDATE/CREATE TABLE:  "
+        + params["database"]
+        + "."
+        + params["env"]
+        + "_"
+        + table_basename
+    )
+    print("##")
+
+    # Check if per_app table already exists
+    table_exists = check_table_exists(session, params, table_basename)
+    print("## table_exists:  ", table_exists)
+
+    # Update the per_app table if an entry does not already exist.
+    success_flag = False
+    for attempt_count in range(3):
+        # NOTE: Retry SQL run 3x for random Athena time-out issue.
```

Comment:
Why 3 times? Is this arbitrary?

Reply:
Based on previous experience of running in to the time-out issue, a re-run had always been successful on a 2nd try. So I believe that if there are 3 failures, it might indicate a different type of issue that needs to be reviewed.

What are your thoughts on adding this as a parameter? This way it could be easily changed without updating the code if needing to tweak the value.

Reply:
A parameter would be great.  I just always hate these types of random solution.  It seems there has to be a repeatable reason for it.  But yeah, QUERY_RETRIES=3